### ISSUE TITLE
Named vs. Unnamed resolution not working as expected

### ISSUE THREAD

**Issue by z4kn4fein** [z4kn4fein/stashbox#63]
Original issue description from #62 by @fernandovm:

**Named vs. Unamed register/resolver:**

I had already tested this approach of register the same single instance with/without name, and yes, I think that is a razonable workaround, but has a undesirable effect, the container return twice the single instance when we invoke the ResolveAll<MyServiceType>() method. Temporarily I'm using distinct extension method to avoiding this behavior.

I think that would be great if we could configure the container globally to deal with named registered instances, in some sense, like as collections, e.g.:

```
container.RegisterInstanceAs<MyServiceType>(instance1, "myService1");
container.RegisterInstanceAs<MyServiceType>(instance2, "myService2");

container.Resolve<MyServiceType>("myService1"); //Returns the instance1 object
container.Resolve<MyServiceType>("myService2"); //Returns the instance2 object
container.Resolve<MyServiceType>(); //Returns the last (or first) object registered
```

_**Note:** Naturally, the use of the distinct to avoid the duplication (caused by duplicated registration workaround) works only for RegisterInstance and Register with Singleton lifetime, when we have just one same object for both the registrations._

**Constructor injection:**

Then, both the suggested approaches works, but in my specific scenario they are a bit intrusive. I need use a minimalistic common dependency injection interface to register/resolve types and objects, because the code that use this interface must be not aware of the dependency injection framework used.

In this case, would be great if we could configure the container globally to use a "[Convention over configuration](https://en.wikipedia.org/wiki/Convention_over_configuration)" approach, the container could resolve the named services using the constructor argument names, maybe a specific Resolver to enable that. 

It is! :)

**Comment by z4kn4fein** [comment]
Hey,

I've pushed a new version to my [pre-release nuget feed](https://www.myget.org/feed/pcsajtai/package/nuget/Stashbox).

It contains 3 new container configuration options:
* **WithUnNamedDependencyResolutionWhenNamedIsNotAvailable()** - it resolves your first issue with the Named vs. Unnamed resolution.
* **TreatParameterOrMemberNamesAsDependencyName()** - this enables the conventional resolution
* **WithMicrosoftExpressionCompiler()** - this one resolves your issue with Xamarin.IOS, I checked and since I introduced the `AssemblyBuilder` and `ModuleBuilder` in my own Expression compiler it doesn't work on those platforms where these types are missing. With this option you can fall back to the Microsoft expression compiler.

Could you please check that the it's working as you expected? 
Let me know if something is still broken. 
Thanks!

**Comment by fernandovm** [comment]
It's amazing, I will try this release and back with feedback!

Thanks!

**Comment by z4kn4fein** [comment]
Hey, 

Happy to hear that! I'll finalize the changes and publish a new version soon!

**Comment by z4kn4fein** [comment]
released [v2.8.4](https://github.com/z4kn4fein/stashbox/releases/tag/2.8.4)

**FYI:** 
I changed the name of the configurations to:
`TreatParameterAndMemberNameAsDependencyName()`
`WithNamedDependencyResolutionForUnNamedRequests()`

**Comment by fernandovm** [comment]
Great, thanks!!!

**Comment by fernandovm** [comment]
Hi Peter,


I did the tests with this version and although the register methods are
working correctly when I call a resolver method it crash the iOS
application.

I repeated the tests with the previous version (pre-release) and the
problem also occurred, I can not explain why I did not realize this in the
previous test. :(


[]'s
FernandoVM

**Comment by z4kn4fein** [comment]
Hey!

:( Could you please send me a bit more detailed description about the issue? Does it throw an exception?

Thanks!

**Comment by fernandovm** [comment]
Hi,

Then, unfortunately the issue is a bit obscure, I not get any error or
exception, or the visual studio freeze or it aborts the debug without no
warnings, silently. This happens exactly when the method 'Resolve<T>()' is
invoked. I'm putting in attach the crash log from iOS device for you see if
help.

[]'s
FernandoVM

**Comment by z4kn4fein** [comment]
Hey,

Unfortunately, the crash dump doesn't contain much useful information, just that it was an access violation. I'll try to reproduce, could you please tell me how you use Stashbox in that project? What features? And what kind of resolution graph do you have? I mean how complex it is?

Thanks!

**Comment by fernandovm** [comment]
Hi Peter,

I'm testing registering/resolvering the classes below:

```
    public class PageResolver : IPageResolver
    {
        private Func<Page> resolver;

        public PageResolver() : this(null)
        {
        }
        public PageResolver(Func<Page> resolver)
        {
            this.resolver = resolver ?? () => null; //omitted
        }

        #region IPageResolver Members

        public Page Current
        {
            get { return this.resolver(); }
        }

        #endregion
    }

    public class DialogServices : IDialogServices
    {
        private Func<Page> resolver;

        public DialogServices(Page current)
        {
            this.resolver = () => current;
        }
        public DialogServices(IPageResolver resolver)
        {
            this.resolver = () => resolver.Current;
        }

        #region IDialogServices Members

        public async Task DisplayAlert(string title, string message, string cancel)
        {
            await this.resolver().DisplayAlert(title, message, cancel);
        }

        public async Task<bool> DisplayAlert(string title, string message, string accept, string cancel)
        {
            return await this.resolver().DisplayAlert(title, message, accept, cancel);
        }

        public async Task<string> DisplayActionSheet(string title, string cancel, string destruction, params string[] buttons)
        {
            return await this.resolver().DisplayActionSheet(title, cancel, destruction, buttons);
        }

        #endregion
    }
```

I can reproduce the problem using the code bellow:

```
        public override bool FinishedLaunching(UIApplication app, NSDictionary options)
        {
            var container = new Stashbox.StashboxContainer(cfg =>
            {
                cfg.WithUnknownTypeResolution()
                    .WithMicrosoftExpressionCompiler()
                    .TreatParameterAndMemberNameAsDependencyName()
                    .WithNamedDependencyResolutionForUnNamedRequests()
                    .WithConstructorSelectionRule(Stashbox.Configuration.Rules.ConstructorSelection.PreferMostParameters);
            });

            container.RegisterInstance(typeof(IPageResolver), new PageResolver(), "pageResolver");
            container.RegisterInstance(typeof(IDialogServices), container.Resolve<Mobile.Services.DialogServices>(), "dialogServices");

            //{omitted}
        }
```

I observed that if I comment the constructor 'public DialogServices(Page current)' I get a resolution exception, if I use the exact code above, the app crashes/aborts.

**Comment by z4kn4fein** [comment]
Thanks for the info, I was able to reproduce, it runs into a stack overflow which is uncatchable by the framework and causes an immediate crash. I'm investigating the fix for this.

**Comment by fernandovm** [comment]
Great news, I'm waiting for news! :))

**Comment by fernandovm** [comment]
Great news, I'm waiting for news! :))

**Comment by z4kn4fein** [comment]
Hey,

Pushed a new version with some fixes to my pre-release feed, could you please give it a try?

Thanks!

**Comment by fernandovm** [comment]
Hi Peter,


I took the test and now everything seems right to me!

Thank you so much!


Best regards,
FernandoVM

**Comment by z4kn4fein** [comment]
Hey,

Cool, happy to hear that!

**Comment by z4kn4fein** [comment]
released [v2.8.5](https://github.com/z4kn4fein/stashbox/releases/tag/2.8.5)

**Comment by fernandovm** [comment]
Great, thanksss!!!!
